### PR TITLE
Faster boolean indexing and indexing by index list

### LINKED ISSUE #2767: Advanced indexing is extremely slow for long dimensions
Motivated by the findings in #2745

```.py
import scipp as sc
import numpy as np
da = sc.data.table_xyz(10_000_000)
```
```.py
%%time
da[da.coords['x'] >= 0.5 * sc.Unit('m')]
```
```
CPU times: user 1min 15s, sys: 1.66 s, total: 1min 16s
Wall time: 6.91 s
```
but with numpy:
```.py
a = da.values
x = da.coords['x'].values
```
```.py
%%time
a[x>=0.5]
```
```
CPU times: user 66.5 ms, sys: 0 ns, total: 66.5 ms
Wall time: 64.1 ms
```

Using index lists shows the same difference between scipp and numpy.

### PR DESCRIPTION
**Pull request by SimonHeybrock**
- Fixes #2767.
- Not equal to NumPy performance, but much better.
- Remove `sc.choose`, was was anyway in an unusable state.

#2759 already brought a big improvement for boolean indexing. In this branch we have (for dimension length of 10 million):

- Bad-case boolean: 300 ms (previously: 550 ms, NumPy: 80 ms).
- Nice-case boolean: 65 ms (NumPy: 50 ms), not much improvement
- Bad case index list: 2 s (previously: 16 s, NumPy: 0.2 s)

NumPy is still much better (especially given that it is single-threaded), but keeping the code generic and simple this may be as much as we can achieve. We can revisit this once someone has actual performance issues with this in practice.

### CODE REVIEW

**Review by jl-wynen — COMMENTED** [review]
It is difficult compare files because a lot of functions moved. Did anything change in the dataset module? Or is `bind_slice_methods.h` the only relevant file?

**Comment by SimonHeybrock** [comment]
There are a lot of changes in the dataset module.

File: lib/dataset/extract.cpp [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    if (i > 0 && values[i - 1] == values[i])
+      continue;    // not an edge
+    if (values[i]) // rising edge
+      indices.emplace_back(i, scipp::size(values));
+    else if (i != 0) // falling edge
+      indices.back().second = i;
+  }
+  return extract_ranges(makeVariable<scipp::index_pair>(Dims{condition.dim()},
+                                                        Shape{indices.size()},
+                                                        Values(indices)),
```
Comment (jl-wynen) [inline lib/dataset/extract.cpp:116]: Can it take advantage of this?
```suggestion
                                                        Values(std::move(indices))),
```

**Review by jl-wynen — APPROVED** [review]
Looks ok as far as I can tell

File: lib/dataset/extract.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply lib/dataset/extract.cpp:116]: No it cannot, since `Variable` is not based on `std::vector`.

Comment (SimonHeybrock) [reply lib/dataset/extract.cpp:116]: We could use `element_array`, but then we would need to do two passes to first determine the size, i.e., more code, i.e., premature optimization.

**Pull request merged**